lsp: Accept text document edits with version zero

mfussenegger opened the pull request:
There were a couple of reports of "Buffer X newer than edits" problems.
We first assumed that it is incorrect for a server to send 0 as a
version - and stated that they should send a `null` instead, given that
in the specification the `textDocument` of a `TextDocumentEdit` is a
`OptionalVersionedTextDocumentIdentifier`.

But it turns out that this was a change in 3.16, and in 3.15 and earlier
versions of the specification it was a `VersionedTextDocumentIdentifier`
and language servers didn't have a better option than sending `0` if
they don't keep track of the version numbers.

So this changes the version check to always accept `0` values.

See

- https://github.com/neovim/neovim/issues/12970
- https://github.com/neovim/neovim/issues/14256
- https://github.com/haskell/haskell-language-server/pull/1727

mfussenegger committed:
lsp: Accept text document edits with version zero

There were a couple of reports of "Buffer X newer than edits" problems.
We first assumed that it is incorrect for a server to send 0 as a
version - and stated that they should send a `null` instead, given that
in the specification the `textDocument` of a `TextDocumentEdit` is a
`OptionalVersionedTextDocumentIdentifier`.

But it turns out that this was a change in 3.16, and in 3.15 and earlier
versions of the specification it was a `VersionedTextDocumentIdentifier`
and language servers didn't have a better option than sending `0` if
they don't keep track of the version numbers.

So this changes the version check to always accept `0` values.

See

- https://github.com/neovim/neovim/issues/12970
- https://github.com/neovim/neovim/issues/14256
- https://github.com/haskell/haskell-language-server/pull/1727

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -470,6 +470,8 @@ function M.apply_text_document_edit(text_document_edit, index)
   -- `VersionedTextDocumentIdentifier`s version may be null
   --  https://microsoft.github.io/language-server-protocol/specification#versionedTextDocumentIdentifier
   if should_check_version and (text_document.version
+      and text_document.version ~= vim.NIL
```
There's a test case that tests for the `vim.NIL` case so I had to add this. Strictly speaking we could probably remove both the test and this given that we normalize `vim.NIL` to `nil` in the rpc layer.

mjlbach replied:
I'm ok removing the test and this in the name of simplification.

mfussenegger replied:
Added a second dedicated commit to remove it, so that the change could be easily reverted individually

mfussenegger committed:
lsp: Remove vim.NIL handling from apply_text_document_edit

The rpc layer normalizes `vim.NIL` to `nil`, so the scenario tested
should never happen.
